src/Polly.Core/Timeout/TimeoutResilienceStrategy.cs

```diff
@@ -0,0 +1,113 @@
+using System;
+using System.Threading.Tasks;
+using Polly.Telemetry;
+
+namespace Polly.Timeout;
+
+internal sealed class TimeoutResilienceStrategy : ResilienceStrategy
+{
+    private readonly TimeProvider _timeProvider;
+    private readonly ResilienceTelemetry _telemetry;
+    private readonly TimeoutStrategy _strategy;
+
+    public TimeoutResilienceStrategy(TimeoutStrategyOptions options, TimeProvider timeProvider, ResilienceTelemetry telemetry)
+    {
+        TimeoutGenerator = options.TimeoutGenerator.CreateHandler();
+        OnTimeout = options.OnTimeout.CreateHandler();
+        _timeProvider = timeProvider;
+        _telemetry = telemetry;
+        _strategy = options.TimeoutStrategy;
+    }
+
+    public Func<TimeoutGeneratorArguments, ValueTask<TimeSpan>> TimeoutGenerator { get; }
+
+    public Func<OnTimeoutArguments, ValueTask> OnTimeout { get; }
+
+    protected internal override async ValueTask<TResult> ExecuteCoreAsync<TResult, TState>(Func<ResilienceContext, TState, ValueTask<TResult>> callback, ResilienceContext context, TState state)
+    {
+        var timeout = await TimeoutGenerator(new TimeoutGeneratorArguments(context)).ConfigureAwait(context.ContinueOnCapturedContext);
+        if (timeout == System.Threading.Timeout.InfiniteTimeSpan || timeout <= TimeSpan.Zero)
+        {
+            // do nothing
+            return await callback(context, state).ConfigureAwait(context.ContinueOnCapturedContext);
+        }
+
+        var previousToken = context.CancellationToken;
+        using var cancellationSource = new CancellationTokenSource();
+        _timeProvider.CancelAfter(cancellationSource, timeout);
+        context.CancellationToken = cancellationSource.Token;
+
+        CancellationTokenRegistration? registration = null;
+
+        if (previousToken.CanBeCanceled)
+        {
+            registration = previousToken.Register(static state => ((CancellationTokenSource)state!).Cancel(), cancellationSource, useSynchronizationContext: false);
+        }
+
+        try
+        {
+            if (_strategy == TimeoutStrategy.Pessimistic)
+            {
+                callback = MakeCancellable(callback);
+            }
+
+            return await callback(context, state).ConfigureAwait(context.ContinueOnCapturedContext);
+        }
+        catch (OperationCanceledException e) when (cancellationSource.IsCancellationRequested && !previousToken.IsCancellationRequested)
+        {
+            _telemetry.Report(TimeoutConstants.OnTimeoutEvent, context);
+
+            await OnTimeout(new OnTimeoutArguments(context, e, timeout)).ConfigureAwait(context.ContinueOnCapturedContext);
+
+            throw new TimeoutRejectedException(
+                $"The timeout resilience strategy did not complete within the timeout of {(int)timeout.TotalSeconds} seconds.",
```

Comment:
I would omit the `int` cast so that if a timeout of, say, 1.5 seconds is used, then the user will see the fractions in the message.

Maybe just format it as a TimeSpan string, rather than convert it to seconds?

Reply:
Done.